### DISCUSSION TITLE
Integration of Langchain create_pandas_dataframe_agent and Prompt Flow

### DISCUSSION THREAD

**Discussion by ahadanjum** [microsoft/promptflow#2071] [Q&A] [answered]
I have a Python file that utilizes AzureChatOpenAI, langchain agents (specifically the create_pandas_dataframe_agent) and Pandas to create an application that allows the user to ask questions based on different SQL tables (loaded as dataframes). My company is preparing to integrate this file into a live web application. Would Prompt Flow be useful for deployment and if so, any tips on how I can use prompt flow to create all the right aspects needed to get it functioning? I have reviewed this page: https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/how-to-integrate-with-langchain?view=azureml-api-2#Overview and was struggling to create the agent and set up the right dependencies, etc

**Comment by brynn-code** [comment]
@jiaochenlu Could you please help on this langchain integration question? Thanks

**Reply by brynn-code** [reply, marked as answer]
And for the deploy option, currently we support deploy flow as an image, or deploy to managed online endpoint from azure machine learning workspace, which kind of deploy option would you prefer to?
Here is the doc of image related deployment guide: https://microsoft.github.io/promptflow/how-to-guides/deploy-a-flow/index.html
And we encourage user deploy from workspace portal if they would like to deploy flow as a managed online endpoint.

**Reply by ahadanjum** [reply]
We will most likely deploy from workspace portal in that case. Thank you for sharing the link and information!

**Comment by jiaochenlu** [comment]
Hi @ahadanjum,  if you have existing Python code for implementing your Langchain agent, you can easily integrate it into a prompt flow by
- Option 1:  Create a flow, add a "Python Tool", place your agent code it into this "Python Tool" node. This offers a straightforward way to incorporate existing code into the flow.
- Option 2:  If you prefer a modular structure for your flow, based on your agent code, you can divide it into distinct segments. Each of these segments can then be implemented in different Python nodes within the flow.

Then to run such a flow in prompt flow, you have to set up the dependencies as mentioning in the documentation:
-  [Prerequisites for environment and runtime](https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/how-to-integrate-with-langchain?view=azureml-api-2#prerequisites-for-environment-and-runtime). Start by setting up a custom environment for your specific Python packages. Then, establish a runtime on your environment.
- [Convert credentials to prompt flow connection](https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/how-to-integrate-with-langchain?view=azureml-api-2#convert-credentials-to-prompt-flow-connection). Assuming you have an Azure OpenAI key stored locally and used by environment variables getting the value in Langchain, however in prompt flow you can create a custom connection to store these keys. You can then configure your Python node to retrieve the key from the custom connection.

**Reply by ahadanjum** [reply]
Hello @jiaochenlu, thank you for providing me with the detailed response. If I go with option 1, would I put all my code under the "@tool" part of the Python tool node? Or should I set up a different tool for each part of the agent code? 

I have uses the requirments.txt file to set up all the correct dependencies I need. Thank you for the help with that

**Reply by brynn-code** [reply]
We may suggest set up different tools for multiple agent, as we may have some system metrics when calling llm when node runs, and it's more clear to manage the process with separate nodes in flow.
